synapse/storage/databases/main/roommember.py

```diff
@@ -1212,6 +1212,30 @@ def _get_forgotten_rooms_for_user_txn(txn: LoggingTransaction) -> Set[str]:
             "get_forgotten_rooms_for_user", _get_forgotten_rooms_for_user_txn
         )
 
+    async def is_locally_forgotten_room(self, room_id: str) -> bool:
+        """Returns whether all local users have forgotten this room_id.
+
+        Args:
+            room_id: The room ID to query.
+
+        Returns:
+            Whether the room is forgotten.
+        """
+
+        sql = """
+            SELECT count(*) FROM local_current_membership
+            INNER JOIN room_memberships USING (room_id, event_id)
+            WHERE
+                room_id = ?
+                AND forgotten = 0;
```

Comment:
Drive-by comment: it might(?) be more performant to `SELECT count(*) > 0` here. That gives the database a chance to stop the query as soon as it finds the first membership that hasn't been forgotten.

(There's also an index `"room_memberships_user_room_forgotten" btree (user_id, room_id) WHERE forgotten = 1` but I'm not sure if there's an easy way to use that here.)

Reply:
https://github.com/matrix-org/synapse/blob/8fa7eb8bcc31bb46bbf3396884fdd97d56160259/synapse/storage/databases/main/roommember.py#L1188-L1207
This looks similar, but is very complicated.

Reply:
I try to create a better sql query. I am feeling in a rabbit hole. It seems that Synapse is not possible to forget a room, at the moment. I think there is a bug in:
https://github.com/matrix-org/synapse/blob/e825f7366b589f61977f63d4fdc598abe052e6a0/synapse/handlers/room_member.py#L1912-L1927
There is a gap of unit tests.

Reply:
Proposal for an other query:
```sql
SELECT room_id, ( 
 SELECT count(*) > 0 FROM room_memberships 
 WHERE room_id = m.room_id AND forgotten = 0 
) AS count 
FROM room_memberships AS m 
WHERE room_id = ? AND forgotten = 1 
GROUP BY room_id;
```

Steps
1. Query if the room is at least one time forgotten - uses partial index on `forgotten = 1`
2. Query this rooms (result of no. 1) that there is no user which has not forgotten (or rejoin) this room

Results
- `count == 0` -> room is forgotten
- `count > 0` -> room is not fully forgotten but at least one time
- no result / `count is None` -> room is not in list (unknown) or room is not at least one time forgotten